kong/clustering/services/negotiation.lua

```diff
@@ -0,0 +1,270 @@
+local DECLARATIVE_EMPTY_CONFIG_HASH = constants.DECLARATIVE_EMPTY_CONFIG_HASH
+local KONG_VERSION
+
+-- it's so annoying that protobuf does not support map to array
+local function wrap_services(service)
+  local wrapped = {}
+  for name, version in pairs(service) do
+    wrapped[name] = {
+      service = version
+    }
+  end
+
+  return wrapped
+end
+
+local function unwrap_services(service)
+  local unwrapped = {}
+  for name, version in pairs(service) do
+    unwrapped[name] = version.service
+  end
+
+  return unwrapped
+end
+
+local _M = {}
+
+local function field_validate(tbl, field, typ)
+  local v = tbl
+  for _, ind in ipairs(field) do
+    if type(v) ~= "table" then
+      error("field cannot be indexed with " .. ind)
+    end
+    v = v[ind]
+  end
+
+  local compare_typ = typ
+  if typ == "array" or typ == "object" then
+    compare_typ = "table"
+  end
+
+  if type(v) ~= compare_typ then
+    local field_name = table_concat(field, '.')
+    error("field \"" .. field_name .. "\" must be of type " .. typ)
+  end
+end
+
+local function verify_request(body)
+  for field, typ in pairs{
+    [{
+      "node",
+    }] = "object",
+    [{
+      "node", "type",
+    }] = "string",
+    [{
+      "node", "version"
+    }] = "string",
+    [{
+      "services_requested"
+    }] = "array",
+  } do
+    field_validate(body, field, typ)
+  end
+end
+
+local function verify_node_compatibility(client_node)
+  if client_node.type ~= "KONG" then
+    error(("unknown node type %q"):format(client_node.type), CLUSTERING_SYNC_STATUS.UNKNOWN)
+  end
+
+  if KONG_VERSION == nil then
+    KONG_VERSION = kong.version
+  end
+
+  local ok, err, result = clustering_utils.check_kong_version_compatibility(KONG_VERSION, client_node.version)
+  if not ok then
+    error(err)
+  end
+  return result
+end
+
+local function negotiate_version(name, versions, known_versions)
+  local versions_set = {}
+  for _, v in ipairs(versions) do
+    versions_set[lower(v.version)] = true
+  end
+
+  for _, v in ipairs(known_versions) do
+    local version = lower(v.version)
+    if versions_set[version] then
+      return v
+    end
+  end
+
+  return { name = name, description = "No valid version" }
+end
+
+local function negotiate_service(name, versions)
+  name = lower(name)
+
+  if type(versions) ~= "table" then
+    error("invalid versions array for service " .. name)
+  end
+
+  local supported_service = supported_services[name]
+  if not supported_service then
+    return {
+      description = "unknown service.",
+    }
+  end
+
+  local service_response = negotiate_version(name, versions, supported_service)
+  return service_response
+end
+
+local function log_negotiation_result(name, version)
+  local ok = version.version ~= nil
+  ngx_log(NOTICE, _log_prefix, "service ",
+    (ok and "accepted" or "rejected"),
+    ": \"", name, "\"",
+    (ok and ", version: " .. version.version or ""),
+    ", ", (ok and "description" or "reason"), ": ", version.description
+  )
```

Comment:
this is clever, but hard to read and worse, hard to search
please do:
```suggestion
  if version.version ~= nil then
    ngx_log(NOTICE, "service accepted: \"", name, "\", version: ", version.version, ", description: ", version.description )
 
  else
    ngx_log(NOTICE, "service rejected: \"", name, "\", reason: ", version.description)
  end
```